lib/python/rose/config_processors/fileinstall.py

```diff
@@ -405,17 +404,15 @@ def _target_install(self, target, conf_tree, work_dir):
                     if not os.path.isfile(target.name):
                         self.manager.fs_util.delete(target.name)
                     handle = open(target.name, "wb")
-                f_bsize = os.statvfs(source.cache).f_bsize
-                source_handle = open(source.cache, 'rb')
-                while True:
-                    bytes_ = source_handle.read(f_bsize)
-                    if not bytes_:
-                        break
-                    try:
-                        handle.write(bytes_.encode())
-                    except AttributeError:
-                        handle.write(bytes_)
-                source_handle.close()
+                async with aiofiles.open(
+                    source.cache, mode='rb'
+                ) as source_handle:
+                    bytes_ = await source_handle.read()
+                # Deal with possibility of different types of handle
+                try:
+                    handle.write(bytes_.encode())
```

Comment:
When you say "Just copy the file" do you mean run a subprocess with cp?

Reply:
That would be one option, yes.

Reply:
This seems just to be a Python copy implementation. After talking with @matthewrmshin we think this might pre-date support for copyfiles in the Python standard library.

I'm a bit worried that people will use Rose to copy huge data files. As this implementation loads the whole file into memory if used with whopping files this could cause swapping or crashes.

Sadly neither `aiofiles` or `asyncio` provide a copy utility, however, as luck would have it `aiofiles` supports reading files in chunks so we can keep the original low-footprint approach.

Here is my Python copy utility using `aiofiles`:

```python
import aiofiles
import asyncio


async def async_copy(src, tgt, chunk_size=1024):
    async with aiofiles.open(tgt, 'bw+') as tgt_handle:
        async with aiofiles.open(src, 'br') as src_handle:
            while True:
                bytes_ = await src_handle.read(chunk_size)
                if not bytes_:
                    break
                await tgt_handle.write(bytes_)


if __name__ == '__main__':
    import sys
    asyncio.run(async_copy(sys.argv[1], sys.argv[2]))
```

Reply:
Thank you @oliver-sanders for bailing me out on this this one. The lesson is to be very wary of `handle = aiofiles.open()` - if not awaited you don't get the correct object back and attempts to read and write will fail. Miserably. Ensure any `handle.close()` is also prefixed by an await.

Reply:
@matthewrmshin one last thing for you to double check...

Two changes:

* files read and written in chunks using asyncio/aiofiles
* files now read in 4098 byte chunks rather than `f_bsize`